lib/route/route.js

```diff
@@ -472,25 +532,122 @@ const checkPageIdParams = (alpha, omega) => {
   return true
 }
 
-const getPreviousPage = (currentPage, userData) => {
+/*
+ *  This is a weird function. I'm not sure how to refactor it (yet) so I'm commenting it up. Hello
+ *
+ *  It does one or both of two things:
+ *
+ *  First, it asks whether or not the current page is a step of an upload journey. If it is: the previous
+ *  page is the parent of the step
+ *
+ *  Second, and if it is not a step of an upload journey, it rewinds and fast-forwards again up to
+ *  the previous page
+ *
+ *  I've tried to do some renaming of variables to make their content a bit clearer. Hope it helps
+ */
+function getPreviousPage (currentPage, userData) {
   const { _id } = currentPage
-  let startPage
-  let instance = getInstance(_id)
-  while (instance._parent) {
-    startPage = instance._parent
-    instance = getInstance(instance._parent)
-  }
 
-  let candidatePage = { _id: startPage }
-  while (candidatePage) {
-    const candidateResultPage = getNextPage(candidatePage, userData, true)
+  /*
+   *  No _id, no previous page
+   */
+  if (_id) {
+    let currentInstance = getInstance(_id)
+
+    /*
+     *  Determine whether the current page is a check or summary step
+     */
+    if (isUploadCheckPage(currentInstance) || isUploadSummaryPage(currentInstance)) {
+      /*
+       *  It is
+       */
+      const { _parent } = currentInstance
+      let candidateInstance = getInstance(_parent)
+
+      /*
+       *  Find the upload page
+       */
+      while (!hasUploadComponents(candidateInstance)) {
+        const { _parent } = candidateInstance
+        candidateInstance = getInstance(_parent)
+      }
+
+      /*
+       *  Found the upload page -- return it
+       */
+      return candidateInstance
+    }
+
+    /*
+     *  Rewind all the way back to the start
+     */
+    let { _parent } = currentInstance
 
-    if (checkPageIdParams(currentPage, candidateResultPage)) {
-      return candidatePage
+    while (Reflect.has(currentInstance, '_parent')) {
```

Comment:
I wrote this and I think it's stupid

I like using `Reflect.has()` and `Reflect.get()` but it's a bit silly when at line 584 I'm destructuring the same object/field pair

I'm going to come back to it

It bothers me that `candidatePage` is just an object with an `_id` at line 594 which is connected to line 584 with this fudge in between and it's very meh